build.rs

```diff
@@ -185,6 +186,41 @@ fn main() -> anyhow::Result<()> {
     .join("\n");
     fs::write(&dest_path, conf_decl).unwrap();
 
+    let metadata = MetadataCommand::new()
+        .manifest_path("Cargo.toml")
+        .features(CargoOpt::AllFeatures)
+        .exec()
+        .unwrap();
+
+    let root_package_id = metadata
+        .root_package()
+        .ok_or(anyhow::anyhow!("Failed to get ID of root package"))?
+        .id
+        .clone();
+
+    let viam_modules : Vec<_> = metadata
+        // Obtain the dependency graph from the metadata and iterate its nodes
+        .resolve.as_ref()
+        .ok_or(anyhow::anyhow!("Dependencies were not resolved"))?
+        .nodes.iter()
+        // Until we find the root node..
+        .find(|node| node.id == root_package_id)
+        .ok_or(anyhow::anyhow!("Root package not found in dependencies"))?
+        // Then iterate the root node's dependencies, selecting only those
+        // that are normal dependencies.
+        .deps.iter()
+        .filter(|dep| dep.dep_kinds.iter().any(|dk| dk.kind == cargo_metadata::DependencyKind::Normal))
+        // And which have a populated `package.metadata.com.viam` section in their Cargo.toml
+        // which has `module = true`
+        .filter(|dep| metadata[&dep.pkg].metadata["com"]["viam"]["module"].as_bool().unwrap_or(false))
+        .collect();
```

Comment:
do you have an idea of what would happen is dependencies also depends on another viam module? would it get populated too? 
If multiple version of the same module exists what's the behavior?

Reply:
So, with the caveat that I don't have much of an understanding of how rust / cargo deal with multiple versions, I think this will work out ok. I deliberately am only selecting direct dependencies of the top level project for auto-registration: transitive dependencies are not automatically registered. So, if you have in the robot project `Cargo.toml`:
```
[dependencies]
my_foo_module = 0.1.0
my_bar_module = 0.1.0
```

And then the `my_foo_module` `Cargo.toml` has
```
[dependencies]
my_bar_module = 0.2.0
my_baz_module = 0.3.0
```

My intention is that the generated project will only call `my_foo_module@0.1.0::register_models` and `my_bar_module@0.1.0::register_models`, since those are directly in the project `Cargo.toml`.

Now, if `my_foo_module` actually wants to handle registration of its dependencies, it can do that in its own `register_models` entry point. In the case above, I'd expect that registering `my_baz_module` would work fine, but an attempt to register `my_bar_module` would fail due to a naming conflict.

Reply:
actually, in the case of naming conflict, we don't return an error because the signature of `register_models` doesn't allow you to return one. This might be a problem actually because the version of bar that would be registered would depend on the ordering in the dependencies section (which I don't think is very intuitive). Should we make a special error enum with a `AlreadyRegistered` option and change the signature of `register_models` to return it (not blocking this PR of course)?

Reply:
Well, in `registry.rs` for the micro-rdk I see this for `register_sensors`:
```
    pub fn register_sensor(
        &mut self,
        model: &'static str,
        constructor: &'static SensorConstructor,
    ) -> Result<(), RegistryError> {
        if self.sensor.contains_key(model) {
            return Err(RegistryError::ModelAlreadyRegistered(model));
        }
        let _ = self.sensor.insert(model, constructor);
        Ok(())
    }
```

So there already is `RegistryError::ModelAlreadyRegistered`, and the generated`register_modules` also returns `anyhow::Result<(), RegistryError>`, and I intend for the generated`$module::register_models` to do so as well, so I do think the error propagates OK, unless I'm misunderstanding you, or rust, which is quite possible.

Regarding order of registrations, it is true that the first one "wins" so the losing one is order dependent. But that's no worse than were things were before: if you were manually editing the `main.rs` of your project to call the registration function for each module, it'd also depend on the order of declaration.